app/controllers/admin/categories_controller.rb

```diff
@@ -0,0 +1,60 @@
+class Admin::CategoriesController < ApplicationController
+  before_action :logged_in_user
+  before_action :is_admin
+
+  def index
+    @categories = Category.paginate(page: params[:page], per_page: 8).order(created_at: :desc)
+  end
+
+  def new
+    @category = Category.new
+  end
+
+  def create
+    @category = Category.new(category_params)
+    if @category.save
+      flash[:success] = "Category was successfully created"
+      redirect_to admin_categories_path
+    else
+      render :new
+    end
+  end
+
+  def edit
+    @category = Category.find(params[:id])
+  end
+
+  def update
+    @category = Category.find(params[:id])
+    if @category.update_attributes(category_params)
+      flash[:success] = "Category was successfully updated"
+      redirect_to admin_categories_path
+    else
+      render :edit
+    end
+  end
+
+  def destroy
+    @category = Category.find(params[:id])
+
+    if @category.posts.count > 0
+      flash[:danger] = 'You can only delete when no more posts belong to this category'
+      return redirect_to admin_categories_path
+    end
+
+    if @category.destroy
+      flash[:success] = 'Category was successfully deleted.'
+      redirect_to admin_categories_path
+    else
+      flash[:danger] = 'Something went wrong'
+      redirect_to admin_categories_path
+    end
+  end
+
+
+  private
+
+  def category_params
+    params.require(:category).permit(:name)
+  end
+end
```

Comment:
I see the `@category = Category.find(params[:id])` is duplicate many times. So you should create a private method `prepare_category` and add a before action for some actions in this controller